src/modules/navigator/navigator_main.cpp

```diff
@@ -902,6 +909,139 @@ Navigator::load_fence_from_file(const char *filename)
+	(void)orb_unadvertise(h);
+}
+
+void Navigator::check_traffic()
+{
+	double lat = get_global_position()->lat;
+	double lon = get_global_position()->lon;
+	float alt = get_global_position()->alt;
+
+	// TODO for non-multirotors predicting the future
+	// position as accurately as possible will become relevant
+	// float vel_n = get_global_position()->vel_n;
+	// float vel_e = get_global_position()->vel_e;
+	// float vel_d = get_global_position()->vel_d;
+
+	bool changed;
+	orb_check(_traffic_sub, &changed);
+
+	float horizontal_separation = 500;
+	float vertical_separation = 500;
+
+	while (changed) {
+
+		transponder_report_s tr;
+		orb_copy(ORB_ID(transponder_report), _traffic_sub, &tr);
+
+		float d_hor, d_vert;
+		get_distance_to_point_global_wgs84(lat, lon, alt,
+						   tr.lat, tr.lon, tr.altitude, &d_hor, &d_vert);
+
+		// predict final altitude (positive is up) in prediction time frame
+		float end_alt = tr.altitude + (d_vert / tr.hor_velocity) * tr.ver_velocity;
+
+		// Predict until the vehicle would have passed this system at its current speed
+		float prediction_distance = d_hor + 1000;
+
+		// If the altitude is not getting close to us, do not calculate
+		// the horizontal separation
+		// since commercial flights do most of the time keep flight levels
+		// check for the current and for the predicted flight level.
+		// we also make the implicit assumption that this system is on the lowest
+		// flight level close to ground in the
+		// (end_alt - horizontal_separation < alt) condition. If this system should
+		// ever be used in normal airspace this implementation would anyway be
+		// inappropriate as it should be replaced with a TCAS compliant solution.
+
+		if ((fabsf(alt - tr.altitude) < vertical_separation) || ((end_alt - horizontal_separation) < alt)) {
+
+			double end_lat, end_lon;
+			waypoint_from_heading_and_distance(tr.lat, tr.lon, tr.heading, prediction_distance, &end_lat, &end_lon);
+
+			struct crosstrack_error_s cr;
+
+			if (!get_distance_to_line(&cr, lat, lon, tr.lat, tr.lon, end_lat, end_lon)) {
+
+				if (!cr.past_end && (fabsf(cr.distance) < horizontal_separation)) {
+
+					// direction of traffic in human-readable 0..360 degree in earth frame
+					int traffic_direction = math::degrees(tr.heading) + 180;
+
+					switch (_param_traffic_avoidance_mode.get()) {
+
+					case 0: {
+							/* ignore */
+							PX4_WARN("TRAFFIC %s, hdg: %d", tr.callsign, traffic_direction);
+							break;
+						}
+
+					case 1: {
+							mavlink_log_critical(&_mavlink_log_pub, "WARNING TRAFFIC %s at heading %d, land immediately", tr.callsign,
+									     traffic_direction);
+							break;
+						}
+
+					case 2: {
+							mavlink_log_critical(&_mavlink_log_pub, "AVOIDING TRAFFIC %s heading %d, returning home", tr.callsign,
+									     traffic_direction);
+
+							// set the return altitude to minimum
+							_rtl.set_return_alt_min(true);
```

Comment:
Let's remove the failsafe action until we get more clarity on what we need to do